### ISSUE TITLE
Constant function should only depend on static variables.

### ISSUE THREAD

**Issue by lrhn** [Pajn/dep-const-function-literals#1]
The specification should also reject const function expressions if they contain references to non-static state. 
That is: if they refer to `this`, any instance member, type variable, or local variable/function declared outside the function, then the function is not a valid const function. 
In effect the function should be closed except for top-level or static content, so that it *could* be converted to a top-level or static function itself, and then referenced by name.

It should probably be a compile-time error when a const function refers to a non-static name, as it currently is for accessing an instance variable from a static member.

**Comment by Pajn** [comment]
>That's too restraining, and counter to how Dart is usually defined.
Yes, I agree. Also the latest meeting notes points out that it has a potential of being very confusing. I will revise the proposal and put the current version as an alternative where these problems are brought up.
https://github.com/dart-lang/dart_enhancement_proposals/blob/master/Meetings/2015-05-20%20DEP%20Committee%20Meeting.md

>Maybe we should consider using "static" instead of "const" as prefix
I understand the concern, however static feels very classy to me. I have to look up if it can be used in this context or if other keywords can be used. Will add the problem to the proposal though.

**Comment by munificent** [comment]
Personally, I think `const` is the right prefix to use for these.

**Comment by lrhn** [comment]
Using `const` is probably more usable.
On the other hand, the C use of `static` in a function-local scope is exactly defining a global variable with a local scope, retaining its value across calls to the function, so it's not completely unprecedented.

**Comment by isaias-b** [comment]
On [stackoverflow.com apeared a question](http://stackoverflow.com/q/35537389/3165552) concerning this issue. I just wanted to place a callback, so that the question's answers can be updated when this is done.
CYA
